### ISSUE TITLE
Trait reform

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust#5527]
**CURRENT STATUS:**

A lot of work has been done and there are only a few remaining tasks.

- [x] #17918: Modify method dispatch to use trait selection machinery, so that multi/conditional-dispatch works.
- [ ] #17657: Generalized where clauses
- [x] #18222: Orphan rules
- [ ] #19470: Coherence rules

**EXAMPLES:**

Here are some issues that contain test cases that ought to work once this is done:

- [x] https://github.com/rust-lang/rust/issues/17594

**ORIGINAL TEXT:**

This is an umbrella issue for collecting together issues with the current trait system.  We need to figure out precisely what the solutions will be and discuss at the retreat, but it's at least useful to know what the *problems* are.

**Comment by aturon** [comment]
Nominating for closure. I don't think there's value in this bug any more.

**Comment by pnkfelix** [comment]
closing; the related work is either done or has been filed as a polish/follow-up ticket
